api/authenticator.py

```diff
@@ -2495,3 +2593,95 @@ def _error_uri(self, redirect_uri, pd):
         )
         params = dict(error=problem_detail_json)
         return redirect_uri + "#" + urllib.urlencode(params)
+
+
+class BaseSAMLAuthenticationProvider(AuthenticationProvider, HasSelfTests):
+    """
+    Base class for SAML authentication providers
+    """
+
+    __metaclass__ = ABCMeta
+
+    NAME = 'SAML 2.0'
+
+    DESCRIPTION = _('SAML 2.0 authentication provider')
+
+    DISPLAY_NAME = NAME
+
+    FLOW_TYPE = 'http://opds-spec.org/auth/saml'
+
+    TOKEN_TYPE = "SAML 2.0 token"
+    TOKEN_DATA_SOURCE_NAME = 'SAML 2.0'
+
+    # After successfully authenticating using SAML protocol patrons receive Bearer tokens used
+    # for confirming their authentication state. This configuration setting defines how long these tokens can be valid
+    SAML_TOKEN_EXPIRATION_DAYS = 'token_expiration_days'
+
+    # This is the default value for that configuration setting.
+    DEFAULT_TOKEN_EXPIRATION_DAYS = 42
```

Comment:
For SAML, authentication can be expired early by logging out or by policies implemented in a given IdP, IdP restarts, etc.

Reply:
Actually, now's a good time to mention that "logging out" of Clever (assuming that were possible) won't invalidate the token that the circulation manager issued to SimplyE. That's because the circulation manager never checks back with Clever after the initial login. If you need to periodically check with the IdP whether the user's session is still valid, that's a new check you'll need to run. I have a vague idea for how to do this using a `Credential` object which I can go into if necessary.

Reply:
Tim is absolutely right, depending on the IdP configuration the session can be very short. SAML assertion contains the following information which can be used to determine how long the SP session be valid:
- AuthnStatement/@SessionNotOnOrAfter
- Conditions/@NotOnOrAfter
- SubjectConfirmationData/@NotOnOrAfter

Please refer to this [page](https://stackoverflow.com/questions/29508906/notonorafter-in-subjectconfirmationdata-and-conditions-and-sessionnotonorafter) for detailed description.

I'm thinking that I could implement the following algorithm by checking these attributes in the following order and taking the first non-empty and setting it as a TTL of a token:
- AuthnStatement/@SessionNotOnOrAfter
- Conditions/@NotOnOrAfter
- SubjectConfirmationData/@NotOnOrAfter

@leonardr, @tdilauro, what do you think?

Reply:
In terms of our system, the 'token' you're talking about would be stored as a `Credential` object. You can set `Credential.expires` to set a TTL.

Then I think it would work like this:
* When someone comes in with a bearer token issued by a CM, you look up their Patron record by a Credential -- the bearer token.
* You then look up their SAML assertion -- a second Credential associated with the same Patron -- incidentally checking whether the assertion has expired.
* If the SAML assertion has expired, the bearer token is treated as invalid, and the patron must go through the login process again.

Reply:
Again, haven't gotten through this PR so not sure what parts of this you've already implemented. I can tell by searching the diff that you're doing _something_ with Credential.

Reply:
@vbessonov Yes, I think it makes sense to use the ...NotOnOrAfter values to determine session `expires` time.

@leonardr The SAML assertion expiration could also be captured in the patron bearer token Credential to avoid the SAML credential lookup (unless something other than the `expires` property is needed from the SAML).

But part of the concern I was trying to get at earlier was that the validity of the assertion can be expired ahead of these times by external events, so the assertion could be invalid even though is hasn't expired.

This statement from the beginning of section 2.5.1.2 of [SAML2 core spec](https://www.oasis-open.org/committees/download.php/35711/sstc-saml-core-errata-2.0-wd-06-diff.pdf) sums it up:

> The NotBefore and NotOnOrAfter attributes specify time limits on the validity of the assertion within the context of its profile(s) of use. They do not guarantee that the statements in the assertion will be correct or accurate throughout the validity period.

I'm not sure if this will be a problem for our profile of use, but just want to make sure that it's in front of both of you now.

Reply:
We need to handle this case:

* Patron has a book out on loan
* Patron asks CM for a link to the book
* CM provides the link
* Patron follows the link, providing the book vendor whatever information is necessary to the vendor to establish credentials (I remain fuzzy on how exactly this step is supposed to work)
* Book vendor checks back with the IdP to validate the patron's authentication
* IdP says "no, they logged out, that session is no longer valid"
* Book vendor sends a 401 in response to the request for the book

In that case, the client needs to know that a 401 _from the book vendor_ means they should try logging in again _to the circulation manager_. The same as if they had gotten a 401 directly from the circulation manager.

Reply:
If a book vendor uses SAML, then IdP will just show a login screen and will ask the patron to enter their credentials again